runtime/onert/core/include/backend/train/ExtraTensorRequest.h

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2024 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __ONERT_BACKEND_EXTRA_TENSOR_REQUEST_H__
+#define __ONERT_BACKEND_EXTRA_TENSOR_REQUEST_H__
+
+#include "backend/train/ExtraTensor.h"
+
+namespace onert
+{
+namespace backend
+{
+namespace train
+{
+
+enum class ExtraTensorLifeTime
+{
+  BACKWARD,            // alive during backward()
+  FORWARD_TO_BACKWARD, // alive from forward to backward()
+};
+class ExtraTensorRequest
+{
+
+public:
+  ExtraTensorRequest(ir::OperandInfo info, ExtraTensorLifeTime lt, ExtraTensor **addr)
```

Comment:
```suggestion
  ExtraTensorRequest(const ir::OperandInfo &info, ExtraTensorLifeTime lt, ExtraTensor **addr)
```

About `ExtraTensor **addr`, can't it be implemented without using a double pointer? If you want to use double pointer, please add `const` to the variable properly.

Reply:
Aha, Thank you for your notice. I'll try to find another way :thinking:

Reply:
Hmm.. I rethink about this. 

Easy way is .. 
* use `ExtraTensor** const` 
* use `ExtraTensor*&` 

Otherwise, `shared_ptr<ExtraTensor>&` looks also possible. 
 

<details> 

About adding `const`, `ExtraTensor**` just becomes `ExtraTensor** const`.
  * `ExtraTensor` should be mutable inside each Layer.
  * `ExtraTensor*` should be mutable, this has to be updated after the extra tensor is registered
  * `ExtraTensor**` can be constant. 
</details>

Reply:
@jyoungyun Could you share your opinion? 

Do you think `ExtraTensor*&` is safe enough? or should we have to find another way?

Reply:
@zetwhite I prefer to use smart pointer instead of raw pointer. The smart pointers do automatic memory management and it lessen tha chances of common pointer-related errors, enhancing code reliability. If you implement this code using `shared_ptr`, how about trying to use it?

Reply:
> If you implement this code using shared_ptr, how about trying to use it? 

Thanks for the suggestion!  

I'm not opposed to using `shared_ptr`. But.. I'm a little worried that **other Tensors** are owned by TensorRegistry (Tensor Registry has `unique_ptr<Tensor>`). When operations access tensors, they access it through a raw pointer. Using `shared_ptr` might change the concept that only TensorRegistry owns the tensors. 

> Using `shared_ptr` might change the concept that only TensorRegistry owns the tensors.  

If changing this isn't a problem, I'd like to try to use `shared_ptr`!

Reply:
@ragmani Could you share your opinion about this? Is it fine to add `shared_ptr<Tensors>` in TensorRegistry?

Reply:
I couldn't find any reason why the double pointer is necessary there. I guess a raw pointer alone is sufficient.

Reply:
I heard from @zetwhite the reason why the double pointer is necessary there.

Reply:
> @ragmani Could you share your opinion about this? Is it fine to add shared_ptr<Tensors> in TensorRegistry?

I couldn't find a way to solve this issue better than using `shared_ptr` or double pointer. And I think it's better to use `shared_ptr&` instead of double raw pointer.

Reply:
@ragmani @jyoungyun I'll update to use `shared_ptr` and re-request reviews soon. Thank you for help!

Reply:
As I understand, double pointer was there since it is output parameter. Is the ownership is shared? If not, we don't need to use `shared_ptr`.

In fact, I am not sure `ExtraTensor` is necessary.

Reply:
The ownership is shared with the `TensorRegistry` if each layer has the ownership.

Reply:
> Is the ownership is shared? 

About this.. IMO there is no concrete answer : 

* Sbd could see that only TensorRegistry takes ownership and Layer borrows(not owns) the tensors.
* Otherwise, Sbd could see that both TensorRegistry and Layer own the tensors.

> In fact, I am not sure ExtraTensor is necessary.

I thought that somehow we need to manage the (Extra or LayerScope) tensors in `core` to plan the memory usage.
So, I tried the way define the tensor as 'extra' and manage it through TensorManager.

Could you share your opinion( or what you are concerned) about this? 
Since this work is not in a hurry, I would like to reflect your opinion.